frontend/src/util/index.ts

```diff
@@ -51,6 +51,88 @@ export function match<T extends string | number, Out>(
         : (arms[value] as (() => Out) | undefined ?? fallback)();
 }
 
+/**
+ * Similar to `match` above, but specifically for use with discriminated unions.
+ * Let's say you have something like
+ *
+ *     type T = { type: "number", value: number } | { type: "string", value: string };
+ *     const t: T = ...;
+ *
+ * Then you can say
+ *
+ *     discriminate(t, "type", {
+ *         number: ({ value }) => ..., // `value` is a `number` here!
+ *         string: ({ value }) => ..., // and a `string` here!
+ *     })
+ */
+export function discriminate<
```

Comment:
Huff, so I'm very torn regarding this.

On the one hand, I love that we can do this in TS. And I see the advantages of "exhaustiveness checking" and being able to use this as expression (which is not possible with the alternatives `if`/`switch`). And I'm usually not scared by some complex signatures and stuff. I like type level shenanigans.

But on the other hand, this is biiiig chunk of non-straight-forward code for very little benefit. At least in this PR it's used only twice. Both times, we do not use exhaustiveness checking and in both cases, the fact that it is an expression doesn't much matter (it's just `return discriminiate(...)`) which can easily be replaced with two `return` statements in the arms. So right now I'd be inclined to say it does not improve the code base overall. 

And honestly, in the interest of time, I'm just not gonna look and try to understand this code. (I mean, at least it's good that you can get the gist from the docs and not worry about the impl.) I'm sure it would be interesting, but yeah. And I worry that if we have to touch it in the future again, it will costs us lots of time. Or other contributors are deterred by this kind of code.

But then again, it is already written now and maybe we will actually use its advantages in the future? Maybe?

So yes, I'm super unsure. I think I might have a very slight preference to remove this, but :shrug:

Reply:
I think I would like to have it. But I'm also biased, of course. x)

And yeah, new contributors are not supposed to look at this impl. I would also say that about `match`. And you can also easily replace most (all?) `match` usages. :shrug: 

It's a bit more complicated than `match`, yes, but it also does cooler stuff, and I kinda hate the idea of using `match` everywhere, but then revert to `switch`/`case` for cases where this pattern could really shine. :thinking:

Reply:
> And you can also easily replace most (all?) `match` usages.

Not sure that's true. `match` is used a lot more and in a lot of expression contexts where you would need a new variable or IIFE to make it work. 

But ok, before we debate this for too long, let's keep it. But as soon as someone runs into problems with it or it needs to be adjusted for some reason, I will again propose to maybe throw it out. But we'll see, maybe it's fine.